programs/bpf_loader/src/syscalls/mod.rs

```diff
@@ -4119,6 +4129,464 @@ mod tests {
+                    sysvar_id_va,
+                    got_clock_va,
+                    offset,
+                    length * 2, // Twice the length.
+                    0,
+                    &mut memory_mapping,
+                )
+                .unwrap(),
+                2,
+            );
+
+            let result = SyscallGetSysvar::rust(
+                &mut invoke_context,
+                sysvar_id_va,
+                got_clock_va,
+                offset,
+                length,
+                0,
+                &mut memory_mapping,
+            );
+            result.unwrap();
+            assert_eq!(got_clock, src_clock);
+
+            let mut clean_clock = create_filled_type::<Clock>(true);
+            clean_clock.slot = src_clock.slot;
+            clean_clock.epoch_start_timestamp = src_clock.epoch_start_timestamp;
+            clean_clock.epoch = src_clock.epoch;
+            clean_clock.leader_schedule_epoch = src_clock.leader_schedule_epoch;
+            clean_clock.unix_timestamp = src_clock.unix_timestamp;
+            assert!(are_bytes_equal(&got_clock, &clean_clock));
+        }
+
+        // Test epoch_schedule sysvar
+        {
+            let sysvar_id_va = 0x100000000;
+            let mut got_epochschedule = EpochSchedule::default();
+            let got_epochschedule_va = 0x200000000;
+            let offset = 0;
+            let length = bincode::serialized_size(&src_epochschedule).unwrap() as u64;
+
+            let mut memory_mapping = MemoryMapping::new(
+                vec![
+                    MemoryRegion::new_readonly(
+                        bytes_of(&sysvar::epoch_schedule::id()),
+                        sysvar_id_va,
+                    ),
+                    MemoryRegion::new_writable(
+                        bytes_of_mut(&mut got_epochschedule),
+                        got_epochschedule_va,
+                    ),
+                ],
+                &config,
+                &SBPFVersion::V2,
+            )
+            .unwrap();
+
+            // `offset + length` greater than the size of the sysvar => `Ok(2)`
+            assert_eq!(
+                SyscallGetSysvar::rust(
+                    &mut invoke_context,
+                    sysvar_id_va,
+                    got_epochschedule_va,
+                    offset,
+                    length * 2, // Twice the length.
+                    0,
+                    &mut memory_mapping,
+                )
+                .unwrap(),
+                2,
+            );
+
+            // Something is missing from the memory mapping step in the syscall...
+            //
+            // For layouts requiring alignment, like `EpochSchedule`
+            // (see `warmup`), the syscall needs to handle the padding using
+            // the memory mapping.
+            // Un-commenting the below test, you'll see failure since `warmup`
+            // will be padded to `8` bytes in the memory map.
```

Comment:
@2501babe ugh this one is rough. We'll need to massage the memory mapping some more for this to work.